library/ssl_tls.c

```diff
@@ -5910,7 +5910,8 @@ static size_t ssl_session_load_tls13( mbedtls_ssl_session *session,
 
     session->key_len = *p++;
 
-    if( session->key_len > (size_t)( end - p ) )
+    if( session->key_len > (size_t)( end - p ) ||
+        session->key_len > sizeof( session->key ) )
```

Comment:
@zhihan Can you explain this additional check? The check in L.5889 accounts for numerous fields as documented, including the key length byte. The key length itself is explicitly checked. I'm missing the overflow potential so far.

Reply:
The check in L.5889 is checking the pointer `p`, rather than the content of the buffer. Here we read the content from the buffer on L.5911. And we immediately check the value we read.

Next we perform a `memcpy`.  The existing check with `end - p` verifies we do not go over the boundary reading the data. The added check verifies we do not go over the boundary on the writing end of the data.

Reply:
Ah, I see - thank you. In this case, I think we should return a different error code - something like `MBEDTLS_ERR_SSL_BUFFER_TOO_SMALL`.

Reply:
Done.